pubsub/streaming-analytics/PubSubToGCS.py

```diff
@@ -0,0 +1,105 @@
+import argparse
+import datetime
+import logging
+
+import apache_beam as beam
+import apache_beam.transforms.window as window
+from apache_beam.options.pipeline_options import PipelineOptions
+
+
+class ProcessDoFn(beam.DoFn):
+    """A distributed processing function that acts on individual elements in
+    the PCollection.
+    """
+
+    def process(self, element, publish_time=beam.DoFn.TimestampParam):
+        """Enrich each Pub/Sub message with its publish timestamps.
+
+        Args:
+            element (bytestring): Pub/Sub message.
+            publish_time: Default to the publish timestamp from Pub/Sub.
+
+        Yields:
+            str: Message body and publish timestamp.
+        """
+        publish_time = datetime.datetime.utcfromtimestamp(
+            float(publish_time)).strftime("%Y-%m-%d %H:%M:%S.%f")
+        yield element.decode('utf-8') + ' published at ' + publish_time
+
+
+class OutputDoFn(beam.DoFn):
+
+    def __init__(self, output_path):
+        self.output_path = output_path
+
+    def process(self, batch, window=beam.DoFn.WindowParam):
+        """Write each batch in its own file to a Google Cloud Storage bucket.
+
+        Args:
+            batch (list of bytestring): list of enriched Pub/Sub messages.
+            window: Window info that has been bound to each batch.
+        """
+
+        ts_format = '%H:%M'
+        window_start = window.start.to_utc_datetime().strftime(ts_format)
+        window_end = window.end.to_utc_datetime().strftime(ts_format)
+        filename = '-'.join([self.output_path, window_start, window_end])
+        contents = '\n'.join(str(element) for element in batch)
```

Comment:
If you do decide to use dictionaries instead of strings, it might make sense to call a `beam.Map` to format it into whatever you like before calling this `DoFn`.